graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/DgsDataLoaderScanningInterceptor.kt

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2024 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.graphql.dgs
+
+import org.dataloader.BatchLoader
+import org.dataloader.BatchLoaderWithContext
+import org.dataloader.MappedBatchLoader
+import org.dataloader.MappedBatchLoaderWithContext
+
+/**
+ * Beans that implement this interface will be called during the component scan
+ * that finds defined @DgsDataLoaders and allows the modification or wrapping of
+ * each DataLoader.
+ *
+ * While this hook appears very similar to [com.netflix.graphql.dgs.DataLoaderInstrumentationExtensionProvider]
+ * there are two important differences:
+ * - DgsDataLoaderScanningInterceptors are called when scanning for @DgsDataLoaders, whereas the
+ *   DataLoaderInstrumentationExtensionProviders are called once per request
+ * - DgsDataLoaderScanningInterceptors are afforded the opportunity to change the type of each DataLoader
+ *   as it is being initially registered. Most notably, this allows for converting BatchLoader and MappedBatchLoader
+ *   into their "WithContext" versions, which can simplify certain types of DataLoader instrumentation.
+ */
+interface DgsDataLoaderScanningInterceptor {
```

Comment:
IMO, a better name might be something like `DataLoaderCustomizer` or `DgsDataLoaderCustomizer`. It is sort of the prevailing pattern already to prefix everything with Dgs, which is a bit unfortunate.

Reply:
Ah good suggestion! And yeah, it's a bummer about the `Dgs` prefix, every time I add a new class I dream of cutting a major version where we just remove it from everything...